WormBase/website
# (2nd attempt) Modernize JS and CSS build

This PR is related to #5435, but with much small "change footprint".

## To test:

See the end result: [http://dev.wormbase.org:9017/](http://dev.wormbase.org:9017/)

To run the code:

* Obtain the code:
In a your website repo (on the dev server):
```
git remote add sibyl https://github.com/sibyl229/website.git
git fetch sibyl 262-js:262-js
git checkout 262-js
```
* Follow the instructions in the README to:
    * [run the catalyst server with pre-build asset](https://github.com/sibyl229/website/tree/262-js#run-the-application-in-development), and
    * [run the catalyst server with live build asset](https://github.com/sibyl229/website/tree/262-js#additional-hightly-recommended-setup-development-enviroment-for-js-and-css-development)

## WARNING
### change to Jenkins is needed before this can be deployed to staging (see below)

## Objectives
To modernize the JavaScript development and build workflow for WormBase Website, such that it works with modern JavaScript UI library (such as React). At the same time, keep current functionalities working.

## Immediate benefit
* be able to use NPM packages (NPM is kinda like CPAN but for JavaScript)
* fix issue with static asset cache not refreshing (i.e. main.min.css => main.[content-hash].css)
* be able to split code into smaller modules (i.e. ES6 module syntax)
* use modern JavaScript syntax (ES6, JSX)
* use modern JavaScript development tools: Webpack, linter, CSS autoprefixer, Jest etc. Refer to [create-react-app](https://github.com/facebookincubator/create-react-app)

## Changes summary
* scaffold `client/` directory with create-react-app, and override relevant configuration with a [forked react-scripts module](https://github.com/facebookincubator/create-react-app/compare/master...sibyl229:master)
* tell Catalyst to serve static assets from Webpack dev server and `build/` as 
appropriate
* have Webpack handle the creation of `root/templates/boilerplate/html` so assets with content hash in their names can be linked to correctly
* load JQuery and JQuery UI from [CDN](http://code.jquery.com)
* for backward compatibility, ensure URL path /css/main.min.css and /js/wormbase.min.js redirects to appropriate `build/` assets

## Changes required for Jenkins

The old commands to uglify/minify CSS or JS need to be removed.
And the following commands will produce the production build of the static assets:

```
cd client/   # every commands below needs to be execute in the `client/` directory
rm -r node_modules/    # removes all installed node modules, just to be safe
npm install   # this re-installs every node module dependencies
nom run build    # this recreate the `client/build/`
```
